src/verbosity.jl

```diff
@@ -0,0 +1,89 @@
+function restore_callsite_source_position!(expr, src)
+    # used to fix the logging source file + line
+    # since we're lowering our verbose logging macros to the
+    # Logging.jl macros; otherwise, they would always report this (verbosity.jl)
+    # file as the logging callsite
+    expr.args[1].args[2] = src
+    return expr
+end
+
+vlogmacrodocs = """
+    @debugv N msg args...
+    @infov N msg args...
+    @warnv N msg args...
+    @errorv N msg args...
+
+"Verbose" logging macros. Drop in replacements of standard logging macros, but an
+additional verbosity level `N` is passed to indicate differing verbosity levels
+for a given log level. The verbosity argument is subtracted from the base log level when passed down
+to the core logging logic, so `@debugv 1 msg` will essentially call `@logmsg Debug-1 msg`.
+
+An `LoggingExtras.LevelOverrideLogger`can then be used to filter on the `level` argument.
+For convenience, the
+[`LoggintExtras.with(f; level, verbosity)`](@ref) function is provided to temporarily
+wrap the current logger with a log level and verbosity subtracted to filter while `f` is executed.
+"""
+
+macro debugv(verbosity::Int, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg (Logging.Debug - $verbosity) $msg $(exs...))),
+        __source__,
+    )
+end
+
+macro infov(verbosity::Int, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg (Logging.Info - $verbosity) $msg $(exs...))),
+        __source__,
+    )
+end
+
+macro warnv(verbosity::Int, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg (Logging.Warn - $verbosity) $msg $(exs...))),
+        __source__,
+    )
+end
+
+macro errorv(verbosity::Int, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg (Logging.Error - $verbosity) $msg $(exs...))),
+        __source__,
+    )
+end
+
+macro logmsgv(verbosity::Int, level, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg ($level - $verbosity) $msg $(exs...))),
+        __source__,
+    )
+end
+
+@eval @doc $vlogmacrodocs :(@logmsgv)
+@eval @doc $vlogmacrodocs :(@debugv)
+@eval @doc $vlogmacrodocs :(@infov)
+@eval @doc $vlogmacrodocs :(@warnv)
+@eval @doc $vlogmacrodocs :(@errorv)
+
+"""
+    LoggingExtras.with(f; level=Info)
+
+Convenience function like `Logging.with_logger` to temporarily wrap
+the current logger with a level filter while `f` is executed.
+That is, the current logger will still be used for actual logging, but
+log messages will first be checked that they meet the `level`
+log level before being passed on to be logged.
+"""
+function with(f; level::Union{Int, LogLevel}=Info, verbosity::Integer=0)
```

Comment:
something to decide before merging is if we want to us something so generic as `with`.
I guess for readability the fact that it has a level argument is enough to hint it is about logging.
But it does seem very generic.

Possible alternative is to make the level a positional argument (and required which I think also might be good)
and make this `with_level`

Something like

```
with_level(Debug; verbosity=3) do 
     foo()
     bar()
end
```

doesn't look aweful to me.
But I won't block the PR over this one way or the other

Reply:
Yeah, I like that better. Just pushed an update so that it's `LoggingExtras.withlevel`.